File: libs/shape-serialization/src/main/kotlin/mono/shape/serialization/ShapeSerializationUtil.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.shape.serialization

import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import mono.graphics.geo.Point

/**
 * A util object for serializing shape to Json and load shape from Json
 */
object ShapeSerializationUtil {
    fun toShapeJson(serializableShape: AbstractSerializableShape): String =
        Json.encodeToString(serializableShape)

    fun fromShapeJson(jsonString: String): AbstractSerializableShape? = try {
        Json.decodeFromString(jsonString)
    } catch (e: Exception) {
        console.error("Error while restoring shapes")
        console.error(e)
        null
    }

    fun toConnectorsJson(connectors: List<SerializableLineConnector>): String =
        Json.encodeToString(connectors)

    fun fromConnectorsJson(jsonString: String): List<SerializableLineConnector> = try {
        Json.decodeFromString(jsonString)
    } catch (e: Exception) {
        console.error("Error while restoring connectors")
        console.error(e)
        emptyList()
    }

    fun toMonoFileJson(name: String, serializableShape: SerializableGroup, offset: Point): String {
        val extra = Extra(name, offset)
        val monoFile = MonoFile(serializableShape, extra)
        return Json.encodeToString(monoFile)
    }

    fun fromMonoFileJson(jsonString: String): MonoFile? = try {
        Json.decodeFromString<MonoFile>(jsonString)
    } catch (e: Exception) {
        // Fallback to version 0
        val shape = fromShapeJson(jsonString) as? SerializableGroup
        if (shape != null) {
            MonoFile(shape, Extra("", Point.ZERO))
        } else {
            null
        }
    }
}

File: monosketch-svelte/src/lib/mono/file/shape-serialization-util.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import { SerializableGroup, AbstractSerializableShape, deserializeShape } from '$mono/shape/serialization/shapes';
import {
    SerializableLineConnector,
    SerializableLineConnectorArraySerializer,
} from '$mono/shape/serialization/connector';
import { Point } from '$libs/graphics-geo/point';
import { MonoFile, Extra } from './mono-file';

export class ShapeSerializationUtil {
    static toShapeJson(serializableShape: AbstractSerializableShape): string {
        // @ts-expect-error toJson is attached by Jsonizable
        return JSON.stringify(serializableShape.toJson());
    }

    static fromShapeJson(jsonString: string): AbstractSerializableShape | null {
        try {
            const json = JSON.parse(jsonString);
            return deserializeShape(json);
        } catch (e) {
            console.error("Error while restoring shapes");
            console.error(e);
            return null;
        }
    }

    static toConnectorsJson(connectors: SerializableLineConnector[]): string {
        return JSON.stringify(SerializableLineConnectorArraySerializer.serialize(connectors));
    }
    /* eslint-disable @typescript-eslint/no-explicit-any */
    static fromConnectorsJson(jsonString: string): SerializableLineConnector[] {
        try {
            const json = JSON.parse(jsonString) as any[];
            return SerializableLineConnectorArraySerializer.deserialize(json);
        } catch (e) {
            console.error("Error while restoring connectors");
            console.error(e);
            return [];
        }
    }
    /* eslint-enable @typescript-eslint/no-explicit-any */

    static toMonoFileJson(monoFile: MonoFile): string {
        // @ts-expect-error toJson is attached by Jsonizable
        const json = monoFile.toJson();
        return JSON.stringify(json);
    }

    static fromMonoFileJson(jsonString: string): MonoFile | null {
        try {
            const json = JSON.parse(jsonString);
            // @ts-expect-error fromJson is attached by Jsonizable
            return MonoFile.fromJson(json);
        } catch (e) {
            // Fallback to version 0
            const shape = this.fromShapeJson(jsonString) as SerializableGroup | null;
            if (shape) {
                return MonoFile.create(shape, [], Extra.create("", Point.ZERO));
            } else {
                return null;
            }
        }
    }
}
